### PR TITLE
Build descriptor of array and field load/store in code

### PR DESCRIPTION
**Pull request by uschindler**
First step: fix array index to adapt type not DEF

This also makes the descriptor of the invokedyanmic in the code. As first step I removed the cast for array index. The operand for storing is not yet adapted - @rmuir will take care

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/LDefArray.java [hunk 1]
```diff
@@ -47,6 +46,7 @@
     ALink analyze(final CompilerSettings settings, final Definition definition, final Variables variables) {
         index.expected = definition.objectType;
```
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/LDefArray.java]: This line is no longer necessary.

**Comment by jdconrad** [comment]
Just one minor comment, but this is great!  Thanks, @uschindler for fixing this up.

**Comment by rmuir** [comment]
Nice step!

for a simple expression like this: `def x = new ArrayList(); return x[0];` you can see the boxing gets removed.

before:
```
ICONST_0
INVOKESTATIC java/lang/Integer.valueOf (I)Ljava/lang/Integer;
INVOKEDYNAMIC arrayLoad(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object; [
 ...
]
```

after:
```
ICONST_0
INVOKEDYNAMIC arrayLoad(Ljava/lang/Object;I)Ljava/lang/Object; [
 ...
]
```

Maybe we should add a small test like this one for method calls, just to be sure if the user passes something crazy that the right stuff happens: https://github.com/uschindler/elasticsearch/blob/8281800b64c6ed06a8858aa46c6c016274bd0520/modules/lang-painless/src/test/java/org/elasticsearch/painless/WhenThingsGoWrongTests.java#L178-L182

**Comment by uschindler** [comment]
I removed the obsolete line.

**Comment by uschindler** [comment]
Should I add a test to something goes wrong?
I would test both arrays and also Lists

**Comment by rmuir** [comment]
Yes, sounds great.

**Comment by uschindler** [comment]
I also have a small change
when building the descriptor, I will change the return value to `after.type`. I also applied this for calls.
To fix return type, we would then only need to change `after`'s type.

**Comment by uschindler** [comment]
OK, I also added test. Should I combine the commits to 1? Otherwise its ready.

**Comment by jdconrad** [comment]
@uschindler Thanks again!  We'll keep the 3 commits.

**Pull request merged**